nullaway/src/main/java/com/uber/nullaway/handlers/contract/fieldcontract/EnsuresNonNullIfHandler.java

```diff
@@ -0,0 +1,303 @@
+                  state,
+                  null));
+    }
+
+    // Clean up state again
+    semanticsHold = false;
+    missingFieldNames = null;
+    methodTreeUnderAnalysis = null;
+    methodSymbolUnderAnalysis = null;
+    returnTreesInMethodUnderAnalysis.clear();
+
+    return true;
+  }
+
+  private void buildUpReturnToEnclosingMethodMap(VisitorState methodState) {
+    returnTreesInMethodUnderAnalysis.clear();
+    new TreePathScanner<Void, Void>() {
+      @Override
+      public Void visitReturn(ReturnTree node, Void unused) {
+        VisitorState returnState = methodState.withPath(getCurrentPath());
+
+        TreePath enclosingMethod =
+            NullabilityUtil.findEnclosingMethodOrLambdaOrInitializer(returnState.getPath());
+        if (enclosingMethod == null) {
+          throw new RuntimeException("no enclosing method, lambda or initializer!");
+        }
+
+        // We only add returns that are directly in the method under analysis
+        if (!enclosingMethod.getLeaf().equals(methodTreeUnderAnalysis)) {
+          return super.visitReturn(node, null);
+        }
+
+        // We add the current return tree to the set as it's in the method
+        returnTreesInMethodUnderAnalysis.add(node);
+        return super.visitReturn(node, null);
+      }
+    }.scan(methodState.getPath(), null);
+  }
+
+  /*
+   * Sub-classes can only strengthen the post-condition.
+   * We check if the list in the child classes is at least the same as in the parent class.
+   */
+  @Override
+  protected void validateOverridingRules(
+      Set<String> overridingFieldNames,
+      NullAway analysis,
+      VisitorState state,
+      MethodTree tree,
+      Symbol.MethodSymbol overriddenMethod) {
+    FieldContractUtils.validateOverridingRules(
+        annotName, overridingFieldNames, analysis, state, tree, overriddenMethod);
+  }
+
+  @Override
+  public void onDataflowVisitReturn(
+      ReturnTree returnTree, NullnessStore thenStore, NullnessStore elseStore) {
+    // We only explore return statements that is inside
+    // the method under validation
+    if (!returnTreesInMethodUnderAnalysis.contains(returnTree)) {
+      return;
+    }
+
+    // We might have already found another return tree that results in
+    // what we need, so we don't keep going deep.
+    if (semanticsHold) {
+      return;
+    }
+
+    Set<String> fieldNames = getAnnotationValueArray(methodSymbolUnderAnalysis, annotName, false);
+    if (fieldNames == null) {
+      throw new RuntimeException("List of field names shouldn't be null");
+    }
+
+    boolean trueIfNonNull = getTrueIfNonNullValue(methodSymbolUnderAnalysis);
+
+    // We extract all the data-flow of the fields found by the
+    // engine in the "then" case (i.e., true case)
+    // and check whether all fields in the annotation parameter are non-null
+    Set<String> nonNullFieldsInPath =
+        thenStore.getReceiverFields(trueIfNonNull ? Nullness.NONNULL : Nullness.NULL).stream()
```

Comment:
I think this works, but it's a little non-obvious.  The dataflow analysis happens to track `Nullness.NULL` at this point, but I'm not sure I would rely on it.  Instead, I would check for `NONNULL` fields in either the `thenStore` or the `elseStore`, depending on `trueIfNonNull`.

Reply:
I now think this logic isn't 100% correct, and that it's important to use `trueIfNonNull` to select the `thenStore` or the `elseStore`.  See comments in tests below.